trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -89,6 +95,85 @@ public BacktranslatedExpression translateExpression(
+	 * @param modifiableGlobals
+	 *            The set of variables listed in the Boogie procedure "modifies" clause
+	 * @return A backtranslated ensures clause with the additional conjuncts
+	 */
+	public BacktranslatedExpression translateEnsuresExpression(
+			final de.uni_freiburg.informatik.ultimate.boogie.ast.Expression expression, final ILocation context,
+			final Set<de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression> modifiableGlobals) {
+		final var intType = new CPrimitive(CPrimitives.INT);
+		final var boolRange = BigInterval.booleanRange();
+
+		if (expression == null) {
+			return computeOldVarEqualities(modifiableGlobals)
+					.map(e -> new BacktranslatedExpression(e, intType, boolRange)).orElse(null);
+		}
+
+		final var mainExpression = translateExpression(expression, context);
+		final var oldVarEqualities = computeOldVarEqualities(modifiableGlobals);
+		if (oldVarEqualities.isPresent()) {
+			return new BacktranslatedExpression(
+					new BinaryExpression(Operator.LOGICAND, mainExpression.getExpression(), oldVarEqualities.get()),
+					new CPrimitive(CPrimitives.INT), BigInterval.booleanRange());
+		}
+
+		return mainExpression;
+	}
+
+	private Optional<Expression> computeOldVarEqualities(
+			final Set<de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression> modifiableGlobals) {
+		final var modifiableNames = modifiableGlobals.stream().map(e -> e.getIdentifier()).collect(Collectors.toSet());
+
+		return mSymbolTable.getGlobalScope().entrySet().stream()
+				.flatMap(e -> makeOldVarEquality(e.getKey(), e.getValue(), modifiableNames).stream())
+				.reduce((eq1, eq2) -> new BinaryExpression(Operator.LOGICAND, eq1, eq2));
+	}
+
+	private Optional<Expression> makeOldVarEquality(final String cId, final SymbolTableValue stv,
+			final Set<String> modifiableNames) {
+		final boolean isTypeDecl = stv.getBoogieDecl() instanceof TypeDeclaration;
+		if (isTypeDecl) {
+			// Type declarations are not global variables
+			return Optional.empty();
+		}
+
+		final var boogieName = stv.getBoogieName();
+		if (modifiableNames.contains(boogieName)) {
+			// No equalities for modifiable variables.
+			return Optional.empty();
+		}
+
+		if (!mMapping.hasVar(boogieName, stv.getDeclarationInformation())) {
+			// We omit the conjunct if the variable cannot be backtranslated and output a warning (sound but imprecise).
+			mReporter.accept("Unknown variable: " + boogieName);
+			return Optional.empty();
+		}
+
+		final var info = mMapping.getVar(boogieName, stv.getDeclarationInformation());
+		final boolean isOnHeap = info.getThird();
+		if (isOnHeap) {
+			// We cannot say whether an on-heap variable is modified, so we omit the conjunct (sound but imprecise).
+			//
+			// TODO We could check if the procedure can modify the heap. To do so, we would need the Boogie names of the
+			// heap data arrays.
+			mReporter.accept("Cannot encode non-modifiability of on-heap variable " + cId);
+			return Optional.empty();
```

Comment:
Can you explain in which case we this branch is entered?

Reply:
Thank you, this makes sense.

>For that we would need access to the heap data arrays -- since you have worked in this backtranslation a lot, do you happen to know if we can easily get them here?

I think it is safe to assume that every heap array starts with `SFO.MEMORY`, so could simply check if any of the `modifiableNames` matches this name.

Reply:
Oh, neat. Implemented in 240548f.